src/Microsoft.AspNetCore.Mvc.TagHelpers/Cache/HybridHtmlFragmentCache.cs

```diff
@@ -0,0 +1,219 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Text;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Html;
+using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Mvc.TagHelpers.Cache
+{
+    /// <summary>
+    /// This implementation of <see cref="IHtmlFragmentCache"/> is able to switch from 
+    /// <see cref="IMemoryCache"/> to <see cref="IDistributedCache"/> registered
+    /// services using the <code>options-distributed</code> attribute.
+    /// </summary>
+    public class HybridHtmlFragmentCache : IHtmlFragmentCache
+    {
+        private const string PriorityOptionName = "priority";
+        private const string DistributedOptionName = "distributed";
+
+        private readonly IMemoryCache _memoryCache;
+        private readonly IServiceProvider _serviceProvider;
+
+        public HybridHtmlFragmentCache(IMemoryCache memoryCache, IServiceProvider serviceProvider)
```

Comment:
No, using the longest matching feature (https://github.com/aspnet/DependencyInjection/blob/dev/src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs#L542). It'll construct it with `IDistributedCache` if the service is registered, otherwise pick the constructor with the single `IMemoryCache` argument.